File: src/main/java/uk/ac/ebi/tsc/tesk/k8s/convert/ExecutorCommandWrapper.java
package uk.ac.ebi.tsc.tesk.k8s.convert;

import org.apache.commons.lang.StringUtils;
import uk.ac.ebi.tsc.tesk.tes.model.TesExecutor;

import java.util.ArrayList;
import java.util.List;
import java.util.StringJoiner;
import java.util.regex.Pattern;

/**
 * Wraps list of executor's command, so that:
 * - if any of executor's stdin/stdout/stderr params is set, the command runs in shell
 * - Each part of the original command (single command/argument) that contained shell special chars is surrounded by single quotes,
 *   plus single quote inside such string are replaced with '"'"' sequence
 * - stdin, stdout, stderr streams are redirected to paths according to executors params
 */
public class ExecutorCommandWrapper {

    private static final Pattern SPECIAL_CHARS = Pattern.compile("[ !\"#$&'()*;<>?\\[\\\\`{|~\\t\\n]");
    private final TesExecutor executor;

    public ExecutorCommandWrapper(TesExecutor executor) {
        this.executor = executor;
    }

    public List<String> getCommandsWithStreamRedirects() {
        List<String> result = new ArrayList<>();
        if (StringUtils.isEmpty(executor.getStdin()) && StringUtils.isEmpty(executor.getStdout()) && StringUtils.isEmpty(executor.getStderr())) {
            return executor.getCommand();
        }
        result.add("/bin/sh");
        result.add("-c");
        StringJoiner sj = new StringJoiner(" ");
        for (String commandPart : executor.getCommand()) {
            if (SPECIAL_CHARS.matcher(commandPart).find()) {
                if (commandPart.contains("'")) {
                    commandPart = commandPart.replaceAll("'", "'\"'\"'");
                }
                commandPart = "'" + commandPart + "'";
            }
            sj.add(commandPart);
        }
        if (!StringUtils.isEmpty(executor.getStdin())) {
            sj.add("<");
            sj.add(executor.getStdin());
        }
        if (!StringUtils.isEmpty(executor.getStdout())) {
            sj.add(">");
            sj.add(executor.getStdout());
        }
        if (!StringUtils.isEmpty(executor.getStderr())) {
            sj.add("2>");
            sj.add(executor.getStderr());
        }
        result.add(sj.toString());
        return result;
    }
}

File: tesk/api/kubernetes/convert/executor_command_wrapper.py
"""Wraps list of executor's command.

Such that:
- If any of executor's stdin/stdout/stderr params is set, the command runs in shell
- Each part of the original command (single command/argument) that contained shell
  special chars is surrounded by single quotes, plus single quote inside such string
  are replaced with '"'"' sequence
- `stdin`, `stdout`, `stderr` streams are redirected to paths according to executors
  params
"""

import re
from typing import List

from tesk.api.ga4gh.tes.models import TesExecutor


class ExecutorCommandWrapper:
    """Wraps executor's command."""

    SPECIAL_CHARS = re.compile(r"[ !\"#$&'()*;<>?\[\\`{|~\t\n]")

    def __init__(self, executor: TesExecutor):
        """Initialize the wrapper."""
        self.executor = executor

    def get_commands_with_stream_redirects(self) -> List[str]:
        """Get command with stream redirects."""
        result = []

        if (
            not self.executor.stdin
            and not self.executor.stdout
            and not self.executor.stderr
        ):
            return self.executor.command

        result.append("/bin/sh")
        result.append("-c")

        command_parts = []

        for command_part in self.executor.command:
            if self.SPECIAL_CHARS.search(command_part):
                # Replace single quotes with '"'"'
                if "'" in command_part:
                    replace_command_part = command_part.replace("'", "'\"'\"'")
                # Quote the command part
                quote_command_part = f"'{replace_command_part}'"
            command_parts.append(quote_command_part)

        if self.executor.stdin:
            command_parts.append("<")
            command_parts.append(self.executor.stdin)

        if self.executor.stdout:
            command_parts.append(">")
            command_parts.append(self.executor.stdout)

        if self.executor.stderr:
            command_parts.append("2>")
            command_parts.append(self.executor.stderr)

        result.append(" ".join(command_parts))

        return result
